### PR TITLE
Pass that removes reshapes post LowerTE

### PR DESCRIPTION
**Pull request by ashutosh-arm**
Introduces a Pass for removing intermediate reshapes post
LowerTE() in AOT compiler. This commit adds pass specific
tests and updates usmp generated workspace pools due to
reduction in number of allocations post reshape removals.

Note: this pass at present does not support first reshape
appearing in the graph. If seen as a useful case, it can be
added in the future.

cc: @manupa-arm @grant-arm

### CODE REVIEW

File: include/tvm/relay/transform.h [hunk 1]
```diff
@@ -580,6 +580,14 @@ TVM_DLL Pass AnnotateUsedMemory();
  */
 TVM_DLL Pass CapturePostDfsIndexInSpans();
 
+/*!
+ * \brief Remove reshapes after lowering the graph.
+ *
+ *
+ * \return The pass.
+ */
+TVM_DLL Pass RemoveReshapes();
```
Comment (manupa-arm) [inline include/tvm/relay/transform.h:589]: Lets rename this to remove StandaloneReshapes

File: src/relay/transforms/remove_reshapes.cc [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+          let_var_to_call_lowered_.Set(var, call_lowered);
+        }
+      }
+    }
+    auto body = this->Mutate(let->body);
+    return WithFields(GetRef<Let>(let), var, value, body);
+  }
+
+  Expr Rewrite_(const CallNode* call, const Expr& post) final {
+    /*
```
Comment (manupa-arm) [inline src/relay/transforms/remove_reshapes.cc:62]: nit : might worth explaining what the reader should get out of this block

File: src/relay/transforms/remove_reshapes.cc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/transform.h>
+
+#include "../op/call/call.h"
+#include "../op/memory/on_device.h"
+
+namespace tvm {
+namespace relay {
+
+TVM_REGISTER_PASS_CONFIG_OPTION("relay.RemoveReshapes", Bool);
```
Comment (manupa-arm) [inline src/relay/transforms/remove_reshapes.cc:34]: Please use something like relay.RemoveStandaloneReshapes.enable or relay.use_remove_standalone_reshapes.
I personally prefer the former as it creates a namespace of any future options of the pass.

**Review by manupa-arm — CHANGES_REQUESTED** [review]
Thanks for this work!
Also worth highlighting in some docs of the pass the value of it and why it should be enabled.

cc : @areusch @altanh @lhutton1

**Review by manupa-arm — CHANGES_REQUESTED** [review]
Thanks @ashutosh-arm. looking good!

I think we need unit tests for the pass as well.
(E.g. https://github.com/apache/tvm/blob/main/tests/python/relay/test_pass_partition_graph.py)

**Review by lhutton1 — COMMENTED** [review]
Great work @ashutosh-arm! Just some small things I picked up on..

File: src/relay/transforms/remove_standalone_reshapes.cc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ * under the License.
+ */
+/*!
+ * \file remove_standalone_reshapes.cc
+ * \brief This file contains the Relay pass for removing unfused reshapes from lowered graph.
+ * InferType() cannot be invoked after calling this pass as it removes reshapes from the call
+ * graph. Many targets only need buffer addresses irrespective of the shapes of them. This makes
+ * reshapes symbolic once the graph has been lowered. Reshape removal results into smaller code
+ * size and reduced buffer allocations. It opens up opportunities of operator fusion in the target
+ * backend. Thus, consequently, it improves the performance of the inference.
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/transforms/remove_standalone_reshapes.cc:26]: I have moved this description to transform.h where it makes more sense.

File: src/relay/transforms/remove_standalone_reshapes.cc [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*!
+ * \file remove_standalone_reshapes.cc
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/transforms/remove_standalone_reshapes.cc:20]: ACK

File: include/tvm/relay/transform.h (hunk 1 above)
Comment (ashutosh-arm) [reply include/tvm/relay/transform.h]: ACK

File: src/relay/transforms/remove_reshapes.cc (hunk 3 above)
Comment (ashutosh-arm) [reply src/relay/transforms/remove_reshapes.cc]: Done

File: src/relay/transforms/remove_reshapes.cc (hunk 2 above)
Comment (ashutosh-arm) [reply src/relay/transforms/remove_reshapes.cc]: Moved this description finally to transform.h

File: src/relay/transforms/remove_standalone_reshapes.cc [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      OnDeviceProps on_device_props = GetOnDeviceProps(on_device_call);
+      if (on_device_props.body.defined() && on_device_props.body->IsInstance<CallNode>()) {
+        const Call call_lowered = Downcast<Call>(on_device_props.body);
+        if (call_lowered.defined() && call_lowered->op.same_as(CallLoweredOp())) {
+          let_var_to_call_lowered_.Set(var, call_lowered);
+        }
+      }
+    }
+    auto body = this->Mutate(let->body);
+    return WithFields(GetRef<Let>(let), var, value, body);
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/transforms/remove_standalone_reshapes.cc:58]: Discussed this offline with @lhutton1  to study the impact of not having ExpandANormalForm() post lowering. It seems the LowerTensorExprMutator takes care of expanding the network during lowering. So, introducing ExpandANormalForm() should not make any difference.

File: src/relay/backend/aot_executor_codegen.cc [hunk 7]
```diff
@@ -1096,6 +1096,12 @@ class AOTExecutorCodegen : public MixedModeVisitor {
           tec::UpdateFunctionMetadata(func, this->function_metadata_, workspace_byte_alignment);
         })(mod);
 
+    transform::PassContext pass_ctx = transform::PassContext::Current();
+    bool enable_remove_reshapes =
+        pass_ctx->GetConfig<Bool>("relay.RemoveStandaloneReshapes.enable", Bool(true)).value();
```
Comment (areusch) [inline src/relay/backend/aot_executor_codegen.cc:1101]: i think can we use snake_case here to match the other pass options?

File: src/relay/transforms/remove_standalone_reshapes.cc [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  /*!  * \brief Returns preceding CallLowered when call is a CallLowered(Reshape) */
```
Comment (areusch) [inline src/relay/transforms/remove_standalone_reshapes.cc:61]: i'm probably missing some context here, but what about just returning the args to reshape()?

**Review by areusch — COMMENTED** [review]
don't have tons of context here, but left a couple suggestions

File: src/relay/transforms/remove_standalone_reshapes.cc (hunk 8 above)
Comment (ashutosh-arm) [reply src/relay/transforms/remove_standalone_reshapes.cc:61]: Graph contains let nodes in between the call_lowered(). I've included the following piece as part of the Rewrite_() as well.

```
    /*
    %1 = call_lowered(@tvmgen_default_non_reshape_function, %input, ...);
    let %x: = on_device(%1, ...);
    %2 = (%x,);
    %3 = call_lowered(@tvmgen_default_fused_reshape, %2, ...,
    "relay_attrs"=__dict__="relay.reshape_only"=1, ...);
    */

```

File: src/relay/backend/aot_executor_codegen.cc (hunk 7 above)
Comment (ashutosh-arm) [reply src/relay/backend/aot_executor_codegen.cc]: ACK

**Review by manupa-arm — APPROVED** [review]
LGTM!

@lhutton1 @areusch maybe have a look ?

**Review by lhutton1 — APPROVED** [review]
LGTM!

File: tests/python/relay/backend/test_pass_remove_standalone_reshapes.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    )
+
+    mod["main"] = MarkReshapeOnlyMutator().visit(mod["main"])
+    mod = RemoveStandaloneReshapes()(mod)
+    reshapes_present = any(["reshape" in gv.name_hint for gv in mod.get_global_vars()])
+    assert reshapes_present, "Reshape should have been removed."
+    return
+
+
+# When reshape layer is the last one in the network
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply tests/python/relay/backend/test_pass_remove_standalone_reshapes.py:100]: Sure, I will make a note of it.

**Comment by lhutton1** [comment]
Thanks @ashutosh-arm @manupa-arm @areusch!

**Pull request merged**